File: library/src/main/java/com/rjsvieira/recyclerviewmultipleviewtypesadapter/DataBindAdapter.java
package com.rjsvieira.recyclerviewmultipleviewtypesadapter;

import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

/**
 * Adapter class for managing a set of data binders
 * <p>
 * Created by yqritc on 2015/03/01, modified by rjsvieira.
 */

public abstract class DataBindAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return getDataBinder(viewType).newViewHolder(parent);
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder viewHolder, int position) {
        int binderPosition = getBinderPosition(position);
        getDataBinder(viewHolder.getItemViewType()).bindViewHolder(viewHolder, binderPosition);
    }

    @Override
    public abstract int getItemCount();

    @Override
    public abstract int getItemViewType(int position);

    protected abstract <T extends DataBinder> T getDataBinder(int viewType);

    protected abstract int getPosition(DataBinder binder, int binderPosition);

    protected abstract int getBinderPosition(int position);

    protected void notifyBinderItemChanged(DataBinder binder, int binderPosition) {
        notifyItemChanged(getPosition(binder, binderPosition));
    }

    protected abstract void notifyBinderItemRangeChanged(DataBinder binder, int positionStart, int itemCount);

    protected void notifyBinderItemInserted(DataBinder binder, int binderPosition) {
        notifyItemInserted(getPosition(binder, binderPosition));
    }

    protected void notifyBinderItemMoved(DataBinder binder, int fromPosition, int toPosition) {
        notifyItemMoved(getPosition(binder, fromPosition), getPosition(binder, toPosition));
    }

    protected abstract void notifyBinderItemRangeInserted(DataBinder binder, int positionStart, int itemCount);

    protected void notifyBinderItemRemoved(DataBinder binder, int binderPosition) {
        notifyItemRemoved(getPosition(binder, binderPosition));
    }

    protected abstract void notifyBinderItemRangeRemoved(DataBinder binder, int positionStart, int itemCount);

    public abstract DataBindAdapter clearDataBinderAdapter();
}

File: library/src/main/java/com/rjsvieira/recyclerviewmultipleviewtypesadapter/DataBindAdapter.kt
package com.rjsvieira.recyclerviewmultipleviewtypesadapter

import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import com.rjsvieira.recyclerviewmultipleviewtypesadapter.views.BaseViewHolder
import java.util.ArrayList
import java.util.HashMap
import kotlin.collections.LinkedHashMap

/**
 * Adapter class for managing a set of data binders
 *
 *
 * Created by yqritc on 2015/03/01, modified by rjsvieira.
 */

open class DataBindAdapter : RecyclerView.Adapter<BaseViewHolder>() {

    private var recyclerView: RecyclerView? = null
    private var binderList: ArrayList<DataBinder<BaseViewHolder>> = ArrayList()
    private var viewTypeMapping: HashMap<Int, Int>? = LinkedHashMap()

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
        this.recyclerView = recyclerView
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
        val position = viewTypeMapping?.get(viewType)!!
        val dataBinder = binderList[position]
        return dataBinder.newViewHolder(parent)
    }

    override fun onBindViewHolder(viewHolder: BaseViewHolder, position: Int) {
        val dataBinder = binderList[position]
        dataBinder.setPosition(position)
        dataBinder.bindViewHolder(viewHolder, position)
    }

    override fun getItemId(position: Int): Long = binderList[position].hashCode().toLong()

    override fun getItemCount(): Int = this.binderList.size

    override fun getItemViewType(position: Int): Int = this.binderList[position].itemViewType

    fun clearRecycledViewPool() {
        this.recyclerView?.recycledViewPool?.clear()
    }

    fun addBinder(binder: DataBinder<BaseViewHolder>?) {
        binder?.let {
            this.binderList.add(binder)
            val position = binderList.indexOf(binder)
            this.viewTypeMapping?.put(binderList[position].itemViewType, position)
            notifyItemInserted(position)
        }
    }

    fun addBinderAtPosition(position: Int, binder: DataBinder<BaseViewHolder>?) {
        binder?.let {
            if (position >= 0) {
                this.binderList.add(position, binder)
                this.viewTypeMapping?.put(binderList[position].itemViewType, position)
                notifyItemInserted(position)
            }
        }
    }

    fun removeBinder(binder: DataBinder<BaseViewHolder>?) {
        binder?.let {
            removeBinderAtPosition(binderList.indexOf(binder))
        }
    }

    fun removeBinderAtPosition(index: Int) {
        if (index >= 0) {
            this.binderList.removeAt(index)
            notifyItemRemoved(index)
            notifyItemRangeChanged(index, itemCount.minus(index).minus(1))
        }
    }

    fun removeBindersInRange(startIndex: Int, endIndex: Int = itemCount.minus(1)) {
        if (startIndex >= 0 && endIndex >= 0 && itemCount > startIndex && itemCount > endIndex) {
            this.binderList.subList(startIndex, itemCount).clear()
            notifyItemRangeRemoved(startIndex, endIndex)
            notifyItemRangeChanged(startIndex, endIndex)
        }
    }

    fun getBinderAtPosition(position: Int): DataBinder<BaseViewHolder>? = if (position < binderList.size) binderList[position] else null

    open fun clearDataBinderAdapter(): DataBindAdapter {
        this.viewTypeMapping?.clear()
        this.binderList.clear()
        return this
    }

}
